### ISSUE TITLE
How to control TCP0 and TCP1 alt modes from MCU

### ISSUE THREAD

**Issue by i509VCB** [LattePandaTeam/LattePanda-Mu#8]
I would like to use the TCP0 and TCP1 as Type-C with alt-mode. Diagram of what I am trying to do:
![image](https://github.com/user-attachments/assets/0e1ab27b-6029-440f-bf33-aed31ddb13d4)

From what I can tell, these will require a PD controller. The full carrier board schematic uses a `IT8851`:
![image](https://github.com/user-attachments/assets/7628c1fd-9a82-4d23-932e-7ac2236dd127)

However I am unable to acquire an IT8851. I will use an MCU and control power delivery from that.

From what I can tell in the [Intel 600 Series Chipset Family PCH datasheet](https://www.intel.com/content/www/us/en/content-details/648364/intel-600-series-chipset-family-platform-controller-hub-pch-datasheet-volume-1-of-2.html), 26.1.1 states the following:

> SMLink1 is used to communicate with USB-C* PD Controller on the platform to
configure different modes such as USB, Display Port, Thunderbolt*, etc. When used
for Integrated USB-C purposes, a soft strap must be set to indicate that integrated
USB-C ports from CPU are being used.

26.2
![image](https://github.com/user-attachments/assets/da6f6c5f-7d6d-488f-826a-a45a106888e7)

This matches with what I see in the schematic for pins 3, 4 and 5. The datasheet also mentions a second I2C port (I2C2 on the module), something with retimers.

Although beyond what the datasheet states, there is nothing describing what messages get sent to and from the PD controller.

Regarding the I2C/Smbus interface, is this defined by the SoC or defined by the BIOS? Is LattePanda able to describe the interface to configure the USB-C port mode and muxing in TCP0 and TCP1 over smbus (what messages, layout of each message)? How many I2C interfaces do I need to use for this (1 or 2)?

I tried to see if coreboot had anything to reference on this. I could not find anything.

**Comment by jamesvici** [comment]
The previously linked datasheet is for a 600-series PCH. The [Intel N100 Datasheet](https://edc.intel.com/content/www/us/en/design/products/platforms/processor-and-core-i3-n-series-datasheet-volume-1-of-2/001/signal-description_19/) lists the pinstrap as follows 
![image](https://github.com/user-attachments/assets/94c5867e-3fd2-4a24-84f6-a281b50a1090)

**Comment by i509VCB** [comment]
> The previously linked datasheet is for a 600-series PCH. The [Intel N100 Datasheet](https://edc.intel.com/content/www/us/en/design/products/platforms/processor-and-core-i3-n-series-datasheet-volume-1-of-2/001/signal-description_19/) lists the pinstrap as follows ![image](https://private-user-images.githubusercontent.com/104049076/350450919-94c5867e-3fd2-4a24-84f6-a281b50a1090.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0MDMxNzYsIm5iZiI6MTcyMTQwMjg3NiwicGF0aCI6Ii8xMDQwNDkwNzYvMzUwNDUwOTE5LTk0YzU4NjdlLTNmZDItNGEyNC04NGY2LWEyODFiNTBhMTA5MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQxNTI3NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YzQyODJiMmI0ZDJlYTlmMGI5ZWYxMGJmNmI1OGMwMzZkMjRmYmQ5MzkxYzZiNTU2NTY0NWMwNzEyMTBjMzc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.0JM--YqOZC88Wi0kEZoElesuNrqrAm9ZVj1KppGAegw)

Yes I forgot to update the original image for the pin strap. Shouldn't change the question still.

**Comment by i509VCB** [comment]
I've discussed this a bit on the Discord, but posting here for reference. CrOS EC does have code to talk to the PD mux on the N100: 

https://github.com/coreboot/chrome-ec/blob/main/zephyr/include/drivers/intel_altmode.h
https://github.com/coreboot/chrome-ec/blob/main/zephyr/drivers/usbc/intel_altmode.c
https://github.com/coreboot/chrome-ec/blob/main/zephyr/subsys/pd_controller/pd_task_intel_altmode.c

These describe the registers which you use to talk to the PD Mux over Type-C. 

I also took a logic analyzer to the full carrier board. I tested with a USB-C to Displayport cable and a USB-C to HDMI + USB hub. Attached are pulseview captures (don't forget the second file for the labels).

[LPMu_SMLink1_USBC_Aug04_2024.zip](https://github.com/user-attachments/files/17469932/LPMu_SMLink1_USBC_Aug04_2024.zip)

The capture does show the `INTEL_ALTMODE_REG_DATA_CONTROL`   and `INTEL_ALTMODE_REG_DATA_STATUS` addresses are used.

This capture only was performed on the `TCP1` port since that is what the full carrier has connected for USB-C. I suspect the address written to the registers, `0x38` for TCP1, states which port to control?
![image](https://github.com/user-attachments/assets/ce1d0c9b-351b-428c-abf3-c254289f1199)
